tests/call_tests.cpp

```diff
@@ -0,0 +1,295 @@
+      input.data = bytes{data.begin(), data.end()};
+
+      auto res = exec(input, callback);
+
+      BOOST_REQUIRE(res);
+      BOOST_REQUIRE(res->action_traces.size() == 1);
+
+      // Since callback information was not provided the result of the
+      // execution is returned in the action return_value
+      auto out = fc::raw::unpack<exec_output>(res->action_traces[0].return_value);
+      BOOST_REQUIRE(out.status == 0);
+
+      BOOST_REQUIRE(out.data.size() >= silkworm::kAddressLength);
+
+      evmc::address result;
+      memcpy(result.bytes, out.data.data()+out.data.size() - silkworm::kAddressLength, silkworm::kAddressLength);
+      return result;
+    }
+};
+
+BOOST_AUTO_TEST_SUITE(call_evm_tests)
+BOOST_FIXTURE_TEST_CASE(call_test_function, call_evm_tester) try {
+  evm_eoa evm1;
+  auto total_fund = intx::uint256(vault_balance("evm"_n));
+  // Fund evm1 address with 100 EOS
+  transfer_token("alice"_n, "evm"_n, make_asset(1000000), evm1.address_0x());
+  auto evm1_balance = evm_balance(evm1);
+  BOOST_REQUIRE(!!evm1_balance);
+  BOOST_REQUIRE(*evm1_balance == intx::exp<256>(10, 18) * 100);
+  total_fund += *evm1_balance;
+
+  // Deploy contract
+  auto token_addr = deploy_test_contract(evm1);
+
+  // Deployment gas fee go to evm vault
+  BOOST_REQUIRE(*evm_balance(evm1) + intx::uint256(vault_balance("evm"_n)) == total_fund);
+
+  // Missing authority
+  BOOST_REQUIRE_EXCEPTION(call_test(token_addr, 1234, "alice"_n, "bob"_n),
+                          missing_auth_exception, eosio::testing::fc_exception_message_starts_with("missing authority"));
+
+  // Account not opened
+  BOOST_REQUIRE_EXCEPTION(call_test(token_addr, 1234, "alice"_n, "alice"_n),
+                          eosio_assert_message_exception, eosio_assert_message_is("caller account has not been opened"));
+
+  // Open
+  open("alice"_n);
+
+  // No sufficient funds in the account so decrementing of balance failed.
+  BOOST_REQUIRE_EXCEPTION(call_test(token_addr, 1234, "alice"_n, "alice"_n),
+                          eosio_assert_message_exception, eosio_assert_message_is("decrementing more than available"));
+
+  // Transfer enough funds
+  transfer_token("alice"_n, "evm"_n, make_asset(1000000), "alice");
+
+  BOOST_REQUIRE(intx::uint256(vault_balance("alice"_n)) == intx::exp<256>(10, 18) * 100);
+  auto total_fund2 = intx::uint256(vault_balance("alice"_n)) + intx::uint256(vault_balance("evm"_n));
+
+  BOOST_REQUIRE_EXCEPTION(call_test(token_addr, 0, "alice"_n, "alice"_n),
+                          eosio_assert_message_exception, eosio_assert_message_is("inline actions must succeed"));
+
+
+  // Call and check results 
+  call_test(token_addr, 1234, "alice"_n, "alice"_n);
+  auto count = get_count(token_addr);
+  BOOST_REQUIRE(count == 1234);
+
+  // Gas go from alice's vault to evm's vault
+  BOOST_REQUIRE(total_fund2 == intx::uint256(vault_balance("alice"_n)) + intx::uint256(vault_balance("evm"_n)));
```

Comment:
You should also check that alice's vault balance has actually decreased. Ideally, you would calculate the gas used and multiply by the gas priced to check for the exact amount it should have been reduced (after gas refund).